File: CS/GridWithFilterPanelButtons/CustomGridControl/CustomGridFilterPanelPainter.cs
using DevExpress.Skins;
using DevExpress.Utils.Drawing;
using DevExpress.XtraGrid.Drawing;
using System;
using System.Drawing;
using System.Linq;

namespace GridWithFilterPanelButtons
{
    public class CustomGridFilterPanelPainter : SkinGridFilterPanelPainter
    {
        public CustomGridFilterPanelPainter(ISkinProvider provider) : base(provider) { }

        protected override Rectangle CalcTextBounds(FilterPanelInfoArgsBase e, Rectangle client)
        {
            CustomGridControlViewInfo viewInfo = (e as CustomGridFilterPanelInfoArgs).viewInfo;
            Rectangle textBounds = base.CalcTextBounds(e, client);
            int customizeButtonWidth = 0, activeButtonWidth = 0;
            if(e.ShowCustomizeButton)
                customizeButtonWidth = viewInfo.GetButtonWidth(e.CustomizeButtonInfo.Bounds);
            if(e.ShowActiveButton)
                activeButtonWidth = viewInfo.GetButtonWidth(e.ActiveButtonInfo.Bounds);
            int textWidth = 0;
            if(e.DisplayText != string.Empty)
            {
                int leftBorder = textBounds.X + textBounds.Width + viewInfo.GetButtonsWidth(viewInfo.LeftButtons);
                int rightBorder = e.Bounds.Width - (viewInfo.GetButtonsWidth(viewInfo.RightButtons) + customizeButtonWidth + activeButtonWidth);
                if(leftBorder > rightBorder)
                {
                    textWidth = e.Bounds.Width - textBounds.X - viewInfo.GetButtonsWidth(viewInfo.LeftButtons) - viewInfo.GetButtonsWidth(viewInfo.RightButtons) - customizeButtonWidth - activeButtonWidth;
                    if(e.AllowMRU) textWidth -= e.MRUButtonInfo.Bounds.Width + viewInfo.Offset;
                    return new Rectangle(textBounds.Location, new Size(textWidth, textBounds.Height));
                }
            }
            return textBounds;
        }
    }
}

File: VB/GridWithFilterPanelButtons/CustomGridControl/CustomGridFilterPanelPainter.vb
Imports DevExpress.Skins
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraGrid.Drawing
Imports System
Imports System.Drawing
Imports System.Linq

Namespace GridWithFilterPanelButtons
    Public Class CustomGridFilterPanelPainter
        Inherits SkinGridFilterPanelPainter

        Public Sub New(ByVal provider As ISkinProvider)
            MyBase.New(provider)
        End Sub

        Protected Overrides Function CalcTextBounds(ByVal e As FilterPanelInfoArgsBase, ByVal client As Rectangle) As Rectangle
            Dim viewInfo As CustomGridControlViewInfo = (TryCast(e, CustomGridFilterPanelInfoArgs)).viewInfo
            Dim textBounds As Rectangle = MyBase.CalcTextBounds(e, client)
            Dim customizeButtonWidth As Integer = 0, activeButtonWidth As Integer = 0
            If e.ShowCustomizeButton Then
                customizeButtonWidth = viewInfo.GetButtonWidth(e.CustomizeButtonInfo.Bounds)
            End If
            If e.ShowActiveButton Then
                activeButtonWidth = viewInfo.GetButtonWidth(e.ActiveButtonInfo.Bounds)
            End If
            Dim textWidth As Integer = 0
            If e.DisplayText <> String.Empty Then
                Dim leftBorder As Integer = textBounds.X + textBounds.Width + viewInfo.GetButtonsWidth(viewInfo.LeftButtons)
                Dim rightBorder As Integer = e.Bounds.Width - (viewInfo.GetButtonsWidth(viewInfo.RightButtons) + customizeButtonWidth + activeButtonWidth)
                If leftBorder > rightBorder Then
                    textWidth = e.Bounds.Width - textBounds.X - viewInfo.GetButtonsWidth(viewInfo.LeftButtons) - viewInfo.GetButtonsWidth(viewInfo.RightButtons) - customizeButtonWidth - activeButtonWidth
                    If e.AllowMRU Then
                        textWidth -= e.MRUButtonInfo.Bounds.Width + viewInfo.Offset
                    End If
                    Return New Rectangle(textBounds.Location, New Size(textWidth, textBounds.Height))
                End If
            End If
            Return textBounds
        End Function
    End Class
End Namespace
